test/rekt/features/broker/topology.go

```diff
@@ -0,0 +1,251 @@
+
+		tOpts := []manifest.CfgFn{triggerresources.WithSubscriber(prober.AsKReference(name), "")}
+
+		if t.delivery != nil {
+			if t.delivery.DeadLetterSink != nil {
+				tOpts = append(tOpts, delivery.WithDeadLetterSink(prober.AsKReference(dlqName), ""))
+			}
+			if t.delivery.Retry != nil {
+				tOpts = append(tOpts, delivery.WithRetry(*t.delivery.Retry, t.delivery.BackoffPolicy, t.delivery.BackoffDelay))
+			}
+		}
+		if t.filter != nil {
+			tOpts = append(tOpts, triggerresources.WithFilter(t.filter.Attributes))
+		}
+
+		triggerName := feature.MakeRandomK8sName(name)
+		f.Setup("Create Trigger"+strconv.Itoa(i)+" with recorder",
+			triggerresources.Install(triggerName, brokerName, tOpts...))
+
+		f.Setup("Trigger"+strconv.Itoa(i)+" is ready",
+			triggerresources.IsReady(triggerName))
+	}
+	return prober
+}
+
+// createExpectedEventMap creates a datastructure for a given test topology created by `createBrokerTriggerDeliveryTopology` function.
+// Things we know from the DeliverySpecs passed in are where failed events from both t1 and t2 should land in.
+// We also know how many events (incoming as well as how many failures the trigger subscriber is supposed to see).
+// Note there are lot of baked assumptions and very tight coupling between this and `createBrokerTriggerDeliveryTopology` function.
+func createExpectedEventPatterns(brokerDS *v1.DeliverySpec, triggers []triggerCfg) map[string]knconf.EventPattern {
+	// By default, assume that nothing gets anything.
+	p := map[string]knconf.EventPattern{
+		"brokerdlq": {
+			Success:  []bool{},
+			Interval: []uint{},
+		},
+	}
+
+	brokerdlq := false
+	for i, t := range triggers {
+		name := fmt.Sprintf("t%d", i)
+		dlqName := fmt.Sprintf("t%ddlq", i)
+		p[name] = knconf.EventPattern{
+			Success:  []bool{},
+			Interval: []uint{},
+		}
+		p[dlqName] = knconf.EventPattern{
+			Success:  []bool{},
+			Interval: []uint{},
+		}
+
+		attempts := knconf.DeliveryAttempts(t.delivery, brokerDS)
+
+		p[name] = knconf.PatternFromEstimates(attempts, t.failCount)
+		if attempts <= t.failCount {
+			if t.delivery != nil && t.delivery.DeadLetterSink != nil {
+				p[dlqName] = knconf.PatternFromEstimates(1, 0)
+			} else {
+				brokerdlq = true
+			}
+		}
+		// TODO: replies do not work yet.
+	}
+
+	if brokerdlq && brokerDS != nil && brokerDS.DeadLetterSink != nil {
+		p["brokerdlq"] = knconf.PatternFromEstimates(1, 0)
+	}
+
+	return p
+}
+
+func assertExpectedRoutedEvents(prober *eventshub.EventProber, expected map[string][]conformanceevent.Event) feature.StepFn {
+	return func(ctx context.Context, t feature.T) {
+		for prefix, want := range expected {
+			got := knconf.ReceivedEventsFromProber(ctx, prober, prefix)
+
+			t.Logf("Expected Events %s; \nGot: %#v\n Want: %#v", prefix, got, want)
+			if len(want) != len(got) {
+				t.Errorf("Wanted %d events, got %d", len(want), len(got))
+			}
+
+			// Check event acceptance.
+			if len(want) != 0 && len(got) != 0 {
```

Comment:
And here.

Reply:
https://github.com/knative/eventing/pull/5296/files/0253d11247321b72acc194108494f195e0e12663#diff-b47aba9e71c3b14be6cbb59b18b050f5482a94eb468b1943c6cd6b2ea3717fb8R169